src/CLIFramework/Extension/DaemonExtension.php

```diff
@@ -0,0 +1,202 @@
+<?php
+namespace CLIFramework\Extension;
+
+use CLIFramework\ServiceContainer;
+use CLIFramework\Command;
+use CLIFramework\Logger;
+use CLIFramework\Exception\ExtensionException;
+use CLIFramework\IO\StreamWriter;
+
+class DaemonExtension extends ExtensionBase
+{
+    private $config;
+    private $logger;
+    private $isNoClose = false;
+    private $isNoChangeDirectory = false;
+    private $command;
+
+    public function __construct()
+    {
+        $container = ServiceContainer::getInstance();
+        $this->config = $container['config'];
+    }
+
+    public static function isAvailable()
+    {
+        return function_exists('pcntl_fork');
+    }
+
+    public function bind(Command $command)
+    {
+        $this->command = $command;
+        $this->bindOptions($command);
+        $this->bindHooks($command);
+    }
+
+    public function run()
+    {
+        if (!$this->isAvailable()) {
+            throw new ExtensionException("pcntl_fork() is not supported.");
+        }
+
+        $this->prepareLogger();
+        $logger = $this->getLogger();
+        $logger->debug('call hook: run.before');
+        $this->callHook('run.before');
+        $logger->debug('call daemonize()');
+        $this->daemonize();
+        $logger->debug('call hook: run.after');
+        $this->callHook('run.after');
+        $logger->debug('isNoClose = ' . var_export($this->isNoClose, true));
+        $logger->debug('isNoChangeDirectory = ' . var_export($this->isNoChangeDirectory, true));
+        $logger->debug('PidFilePath = ' . $this->getPidFilePath());
+    }
+
+    public function noClose()
+    {
+        $this->isNoClose = true;
+    }
+
+    public function noChangeDirectory()
+    {
+        $this->isNoChangeDirectory = true;
+    }
+
+    public function getPidFilePath()
+    {
+        if ($this->getCommandOptions() && $this->getCommandOptions()->{'pid-file'}) {
+            return $this->getCommandOptions()->{'pid-file'};
+        }
+        $pid = getmypid();
+        $pidFile = $this->command ? $this->command->getName() : $pid;
+        return $this->config->getPidDirectory() . "/$pidFile.pid";
+    }
+
+    private function bindHooks($command)
+    {
+        $extension = $this;
+        $command->addHook('execute.before', function() use ($extension) {
+            $extension->run();
+        });
+        $command->addHook('execute.after', function() use ($extension) {
+            @unlink($extension->getPidFilePath());
+        });
+    }
+
+    private function bindOptions($command)
+    {
+        $options = $command->getOptionCollection();
+        if (!$options) {
+            return;
+        }
+        $options->add('pid-file?', 'The path of pid file.');
```

Comment:
```php
if ($options = $command->getOptionCollection()) {
    $options->add( .... );
}
```

but I think `getOptionCollection` should always return an option collection? no?